File: src/core/hook.ts
import * as os from 'os';
import * as path from 'path';
import { spawnSync } from 'child_process';
import { rewriteCommand } from './registry';
import { readFileIfExists, safeJsonParse } from './utils';

// Node-free Claude Code PreToolUse hook. Registered in settings.json as
// `tok hook claude`: Claude Code pipes the tool-call JSON to stdin and reads our
// decision JSON from stdout. Doing the whole protocol inside tok (no shell script,
// no `node -e`, no `jq`) is what lets tok ship as a single standalone binary.
//
// Returns the JSON string to print, or null to pass the command through untouched.
export function buildClaudeHookOutput(payload: string): string | null {
  let obj: { tool_name?: unknown; tool_input?: { command?: unknown } } | null;
  try {
    obj = JSON.parse(payload);
  } catch {
    return null;
  }
  if (!obj || String(obj.tool_name || '') !== 'Bash') return null;
  const command = String(obj.tool_input?.command || '');
  if (!command) return null;

  const outcome = rewriteCommand(command);
  if (outcome.kind !== 'allow' && outcome.kind !== 'ask') {
    return null; // none / deny → leave the command untouched
  }

  const toolInput = { ...(obj.tool_input || {}), command: outcome.rewritten };
  const hookSpecificOutput: Record<string, unknown> = {
    hookEventName: 'PreToolUse',
    updatedInput: toolInput,
  };
  // "allow" auto-approves the rewritten command; "ask" leaves the user prompt intact.
  if (outcome.kind === 'allow') {
    hookSpecificOutput.permissionDecision = 'allow';
    hookSpecificOutput.permissionDecisionReason = 'tok auto-rewrite';
  }
  return JSON.stringify({ hookSpecificOutput });
}

// How the hook should invoke tok. Preference:
//   1. `tok` on PATH (a global npm link or a binary already on PATH) — clean + portable.
//   2. A packaged single-file binary that isn't on PATH yet (e.g. mid-install): invoke
//      ourselves by absolute path so the hook works before PATH changes take effect.
//   3. A source checkout: `node <abs main.js>`.
export function resolveTokInvocation(): string {
  if (whichTok()) return 'tok';
  if ((process as unknown as { pkg?: unknown }).pkg) {
    const exe = process.execPath.replace(/\\/g, '/');
    return /\s/.test(exe) ? `"${exe}"` : exe;
  }
  const mainJs = path.resolve(__dirname, '..', 'main.js').replace(/\\/g, '/');
  return `node ${mainJs}`;
}

function whichTok(): boolean {
  const isWin = process.platform === 'win32';
  const r = spawnSync(isWin ? 'where' : 'which', ['tok'], { encoding: 'utf8', shell: isWin });
  return r.status === 0;
}

// The command string registered in settings.json (e.g. "tok hook claude").
export function claudeHookCommand(): string {
  return `${resolveTokInvocation()} hook claude`;
}

// Read the tok PreToolUse hook command currently registered in ~/.claude/settings.json
// (matches both the command form and the legacy tok-rewrite.sh script).
export function readRegisteredClaudeCommand(): string | null {
  const p = path.join(os.homedir(), '.claude', 'settings.json');
  const raw = readFileIfExists(p);
  if (!raw) return null;
  const cfg = safeJsonParse<{ hooks?: { PreToolUse?: Array<{ hooks?: Array<{ command?: string }> }> } }>(raw);
  const pre = cfg?.hooks?.PreToolUse;
  if (!Array.isArray(pre)) return null;
  for (const entry of pre) {
    for (const h of entry.hooks || []) {
      const cmd = String(h.command || '');
      if (/hook\s+claude/.test(cmd) || /tok-rewrite\.sh/.test(cmd)) return cmd;
    }
  }
  return null;
}

// Confirm the hook rewrites a fake Bash tool-call to a `tok` command. This runs the
// exact function the registered `tok hook claude` command executes when Claude Code
// fires it — in-process, so the self-check is reliable. (We deliberately do NOT
// re-spawn tok here: a packaged binary spawning itself with piped stdin is unreliable
// under pkg, and Claude Code — regular Node — invokes the hook fine regardless.) The
// wiring around it (command registered, `tok` on PATH) is checked separately.
export function probeClaudeHook(_command?: string): { pass: boolean; rewrite?: string; reason?: string } {
  const payload = JSON.stringify({ tool_name: 'Bash', tool_input: { command: 'git status' } });
  const out = buildClaudeHookOutput(payload);
  if (!out) return { pass: false, reason: 'hook did not rewrite a Bash git command' };
  const parsed = safeJsonParse<{ hookSpecificOutput?: { updatedInput?: { command?: unknown } } }>(out);
  const rewritten = parsed?.hookSpecificOutput?.updatedInput?.command;
  if (typeof rewritten === 'string' && rewritten.startsWith('tok ')) return { pass: true, rewrite: rewritten };
  return { pass: false, reason: 'unexpected hook output' };
}

File: internal/hook/hook.go
// Package hook implements the Node-free `tok hook claude` PreToolUse hook: Claude Code
// pipes the tool-call JSON to stdin and reads the rewrite decision from stdout. Doing the
// whole protocol inside tok (no shell script, no jq) is what lets tok ship as one binary.
package hook

import (
	"bytes"
	"encoding/json"
	"os"
	"os/exec"
	"path/filepath"
	"regexp"
	"runtime"
	"strings"

	"github.com/Kevalgor12/tok-proxy/internal/registry"
	"github.com/Kevalgor12/tok-proxy/internal/util"
)

// BuildClaudeHookOutput returns the JSON to print for a PreToolUse payload, and whether
// there is anything to print (false = pass the command through untouched).
func BuildClaudeHookOutput(payload string) (string, bool) {
	var obj struct {
		ToolName  string         `json:"tool_name"`
		ToolInput map[string]any `json:"tool_input"`
	}
	if json.Unmarshal([]byte(payload), &obj) != nil {
		return "", false
	}
	if obj.ToolName != "Bash" {
		return "", false
	}
	command, _ := obj.ToolInput["command"].(string)
	if command == "" {
		return "", false
	}

	outcome := registry.RewriteCommand(command)
	if outcome.Kind != "allow" && outcome.Kind != "ask" {
		return "", false // none / deny leaves the command untouched
	}

	// Carry the original tool_input over, overriding just the command.
	toolInput := map[string]any{}
	for k, v := range obj.ToolInput {
		toolInput[k] = v
	}
	toolInput["command"] = outcome.Rewritten

	hookSpecificOutput := map[string]any{
		"hookEventName": "PreToolUse",
		"updatedInput":  toolInput,
	}
	if outcome.Kind == "allow" {
		hookSpecificOutput["permissionDecision"] = "allow"
		hookSpecificOutput["permissionDecisionReason"] = "tok auto-rewrite"
	}

	out, ok := marshalNoEscape(map[string]any{"hookSpecificOutput": hookSpecificOutput})
	if !ok {
		return "", false
	}
	return out, true
}

// marshalNoEscape matches JSON.stringify: no HTML escaping, no trailing newline.
func marshalNoEscape(v any) (string, bool) {
	var buf bytes.Buffer
	enc := json.NewEncoder(&buf)
	enc.SetEscapeHTML(false)
	if err := enc.Encode(v); err != nil {
		return "", false
	}
	return strings.TrimRight(buf.String(), "\n"), true
}

// ResolveTokInvocation is how the hook should call tok - always an absolute path, never
// bare `tok`. Claude Code runs the hook in a fresh shell (Git Bash on Windows) whose PATH
// may not include tok's directory, especially under Store/MSIX-packaged hosts. os.Executable
// is the running binary's absolute path, so the hook works regardless of PATH.
func ResolveTokInvocation() string {
	if exe, err := os.Executable(); err == nil && exe != "" {
		if abs, err := filepath.Abs(exe); err == nil {
			exe = abs
		}
		return shellPath(exe)
	}
	if p := whichTokPath(); p != "" {
		return shellPath(p)
	}
	return "tok"
}

// ClaudeHookCommand is the command string registered in settings.json.
func ClaudeHookCommand() string { return ResolveTokInvocation() + " hook claude" }

func shellPath(p string) string {
	s := strings.ReplaceAll(p, `\`, "/")
	if strings.ContainsAny(s, " \t") {
		return `"` + s + `"`
	}
	return s
}

func whichTokPath() string {
	prog := "which"
	if runtime.GOOS == "windows" {
		prog = "where"
	}
	out, err := exec.Command(prog, "tok").Output()
	if err != nil {
		return ""
	}
	for _, line := range strings.Split(string(out), "\n") {
		if line = strings.TrimSpace(line); line != "" {
			return line
		}
	}
	return ""
}

var registeredCmdRe = regexp.MustCompile(`hook\s+claude|tok-rewrite\.sh`)

// ReadRegisteredClaudeCommand returns the tok PreToolUse command in ~/.claude/settings.json
// (both the command form and the legacy shell script), or "" if none.
func ReadRegisteredClaudeCommand() string {
	home, err := os.UserHomeDir()
	if err != nil {
		return ""
	}
	raw, ok := util.ReadFileIfExists(filepath.Join(home, ".claude", "settings.json"))
	if !ok {
		return ""
	}
	var cfg struct {
		Hooks struct {
			PreToolUse []struct {
				Hooks []struct {
					Command string `json:"command"`
				} `json:"hooks"`
			} `json:"PreToolUse"`
		} `json:"hooks"`
	}
	if json.Unmarshal([]byte(raw), &cfg) != nil {
		return ""
	}
	for _, entry := range cfg.Hooks.PreToolUse {
		for _, h := range entry.Hooks {
			if registeredCmdRe.MatchString(h.Command) {
				return h.Command
			}
		}
	}
	return ""
}

// ProbeClaudeHook runs the hook's decision logic on a fake `git status` payload and
// confirms it rewrites to a tok command. In-process, so the self-check is reliable
// (a packaged binary re-spawning itself with piped stdin is not).
func ProbeClaudeHook() (pass bool, rewrite, reason string) {
	out, ok := BuildClaudeHookOutput(`{"tool_name":"Bash","tool_input":{"command":"git status"}}`)
	if !ok {
		return false, "", "hook did not rewrite a Bash git command"
	}
	var parsed struct {
		HookSpecificOutput struct {
			UpdatedInput struct {
				Command string `json:"command"`
			} `json:"updatedInput"`
		} `json:"hookSpecificOutput"`
	}
	if json.Unmarshal([]byte(out), &parsed) != nil {
		return false, "", "unexpected hook output"
	}
	if rw := parsed.HookSpecificOutput.UpdatedInput.Command; strings.HasPrefix(rw, "tok ") {
		return true, rw, ""
	}
	return false, "", "unexpected hook output"
}
